### PR TITLE
Draft API specification.

### PR DESCRIPTION
**Pull request by gsaluja9**
**Starting to implement an Object mapper API.**
It would enable the users to interact with athena using an API that alleviates the need to be familiar with the [native API](https://docs.aperturedata.io/index.html).

It is going to be designed on the guidelines of a similar ORM library, for example [SQL Alchemy](https://www.sqlalchemy.org/) or [Django's ORM](https://docs.djangoproject.com/en/4.0/topics/db/queries/).

The following are some considerations for implementing it this time:
- It will support CRUD operations for the object types covered by the native API.
- It will treat the Entities in a more or less common way, and implement the basic DB operations (rather than the prescribed ways of actions on these objects.
- The main goal is will be to add simplicity and intuitiveness for interacting with VDMS. The current API requires the users to familiarize themselves with the API.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -132,3 +132,7 @@ dmypy.json
 *.adb.csv
 *.jpg
 *.npy
+
+env3/*
+test/aperturedb/db/*
+test/input/blobs/*
```
Comment (luisremis) [inline .gitignore:138]:
```suggestion
test/input/blobs/*

```

File: README.md [hunk 2]
```diff
@@ -13,3 +13,65 @@ elements in ApertureDB at the object level.
 * NotebookHelpers.py provides helpers to show images/bounding boxes on Jupyter Notebooks
 
 For more information, visit https://aperturedata.io
+
+
+## Object Mapper API
+This API allows the user to interact with the database with the idea of objects that can be Created, Read, Updated and Deleted through an intutive interface involving these objects.
```
Comment (luisremis) [inline README.md:19]:
```suggestion
This API allows the user to interact with the database with the idea of objects that can be 
Created, Read, Updated and Deleted through an intuitive interface involving these objects.
```

File: README.md [hunk 3]
```diff
@@ -13,3 +13,65 @@ elements in ApertureDB at the object level.
...
+
+It supports the following types of Objects in the Database as Python objects:
+| Object | Dcumentation |
```
Comment (luisremis) [inline README.md:22]:
```suggestion
| Object | Documentation |
```

File: README.md [hunk 4]
```diff
@@ -13,3 +13,65 @@ elements in ApertureDB at the object level.
...
+| BoundingBox || 
+| Descriptor||
+| DescriptorSet||
+| Entity||
+| Image||
+| Video||
+
+
+### Examples
+**Create an Image**
```
Comment (luisremis) [inline README.md:34]: this is a great start, but what we would like is two separate sets of documentation: 
1) docstrings or similar, to let users understand the API. Ideally, we would like to generate an html documentation (like this: https://pytorch.org/docs/stable/torch.html (which actually uses read-the-docs))
2) Example documentation, for users to see some use-cases. For the examples, you can consider using read-the-docs/sphinx as we are using for documentation ApertureDB API

File: README.md [hunk 5]
```diff
@@ -13,3 +13,65 @@ elements in ApertureDB at the object level.
...
+
+```
+from aperturedb import Connector, Images
+
+db_connector = Connector.Connector(user="admin", password="admin")
+images = Images.Images(db_connector)
+with open(abs_path, "rb") as instream:
+    img_data = instream.read()
```
Comment (luisremis) [inline README.md:42]: where is img_data used?

File: README.md [hunk 6]
```diff
@@ -13,3 +13,65 @@ elements in ApertureDB at the object level.
...
+```
+
+**Connect an Image to BoundingBox**
+```
+from aperturedb import Connector, Images, BoundingBoxes
+
+db_connector = Connector.Connector(user="admin", password="admin")
+
+bboxes = BoundingBoxes.BoundingBoxes(db_connector)
+bbox = bboxes.get(id=34)
```
Comment (luisremis) [inline README.md:71]: this is a good start. Are there other, more friendly, ways of interacting with the objects without having to specify ids? I would try to come up with something higher level than that.

File: aperturedb/BoundingBoxes.py [hunk 7]
```diff
@@ -0,0 +1,7 @@
+from aperturedb.Repository import Repository
+
+
+class BoundingBoxes(Repository):
+    def __init__(self, db, batch_size=100):
+        super().__init__(db)
+        self._object_type = "BoundingBox"
```
Comment (luisremis) [inline aperturedb/BoundingBoxes.py:7]:
```suggestion
        self._object_type = "BoundingBox"

```

File: test/test_Image.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    def test_enumerate(self):
+        for image in self.images.filter():
+            self.assertIsNotNone(image)
+            self.assertTrue(hasattr(image, "_uniqueid"))
+
+        
+
+        
+if __name__ == '__main__':
+    unittest.main()
```
Comment (luisremis) [inline test/test_Image.py:107]:
```suggestion
    unittest.main()

```

File: test/utils.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    images = Images.Images(db)
+    entities = Repository.Repository(db)
+    bboxes = BoundingBoxes(db)
+    print("Cleaning up DB")
+    for image in images.filter():
+        image.delete()
+    for entity in entities.filter():
+        entity.delete()
+    for bbox in bboxes.filter():
+        bbox.delete()
```
Comment (luisremis) [inline test/utils.py:19]:
```suggestion
        bbox.delete()

```

File: test/utils.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
+from aperturedb import Images, Repository
+from aperturedb.BoundingBoxes import BoundingBoxes
+
+
+    
+def cleanDB(db):
```
Comment (luisremis) [inline test/utils.py:6]: scary name for a method.

**Review by luisremis — CHANGES_REQUESTED** [review]
This is going in the right direction. Before reviewing any implementation details, it would be nice to have a more complete description of the API, where most of the discussion will go.

**Comment by gsaluja9** [comment]
done in another.

**Pull request closed without merging**